### PR TITLE
Provides the initial implementation of the Tycho Metrics Dashboard.

### PR DESCRIPTION
**Pull request by andyHa**

### CODE REVIEW

File: src/main/java/sirius/biz/analytics/charts/Dataset.java [hunk 1]
```diff
@@ -63,6 +63,17 @@ public Dataset addValue(Number value) {
         return this;
     }
 
+    /**
+     * Adds the given values to the dataset
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/charts/Dataset.java:67]:
```suggestion
     * Adds the given values to the dataset.
```

File: src/main/java/sirius/biz/analytics/charts/Dataset.java [hunk 2]
```diff
@@ -63,6 +63,17 @@ public Dataset addValue(Number value) {
...
+     *
+     * @param values the values to add
+     * @return the dataset itself for fluent method calls
+     */
+    public Dataset addValues(List<Integer> values) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/charts/Dataset.java:72]: In the addValue method a Number is passed in, here it is an Integer. Shouldn't this be more consistent?

File: src/main/java/sirius/biz/analytics/metrics/MetricQuery.java [hunk 3]
```diff
@@ -272,6 +343,20 @@ private int determineLimit(Interval interval) {
         };
     }
 
+    /**
+     * Determines  the default limit (e.g. for {@link #valuesUntil(LocalDate, int)}.
```
Comment (sabieber) [inline src/main/java/sirius/biz/analytics/metrics/MetricQuery.java:347]:
```suggestion
     * Determines the default limit (e.g. for {@link #valuesUntil(LocalDate, int)}.
```

File: src/main/java/sirius/biz/tycho/dashboards/BaseTimeseriesChart.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Provides basic functionality needed by all time-series based charts.
+ */
+public abstract class BaseTimeseriesChart implements Chart {
+
+    private List<String> labels = Collections.emptyList();
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/dashboards/BaseTimeseriesChart.java:26]: Shouldn't this be mutable and final and be modified/filled in the methods below?

File: src/main/java/sirius/biz/tycho/dashboards/BaseTimeseriesChart.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+    private List<Dataset> datasets = new ArrayList<>();
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/dashboards/BaseTimeseriesChart.java:27]: Can be final?

File: src/main/java/sirius/biz/tycho/dashboards/BasicChartProvider.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    private Set<String> permissions;
+
+    @Override
+    public boolean isAccessible() {
+        if (permissions == null) {
+            permissions = Permissions.computePermissionsFromAnnotations(getClass());
+        }
+
+        UserInfo user = UserContext.getCurrentUser();
+        return permissions.stream().allMatch(user::hasPermission);
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/dashboards/BasicChartProvider.java:32]: `UserInfo::hasPermissions` could be used here

File: src/main/java/sirius/biz/tycho/dashboards/BasicKeyMetricProvider.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    private Set<String> permissions;
+
+    @Override
+    public boolean isAccessible() {
+        if (permissions == null) {
+            permissions = Permissions.computePermissionsFromAnnotations(getClass());
+        }
+
+        UserInfo user = UserContext.getCurrentUser();
+        return permissions.stream().allMatch(user::hasPermission);
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/dashboards/BasicKeyMetricProvider.java:37]: s.a.

File: src/main/java/sirius/biz/tycho/dashboards/Chart.java [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+ * Represents a chart which is generated by {@link ChartProvider#resolveChart(String, String)}.
+ */
+public interface Chart {
+
+    /**
+     * Generates the JSON representation required by the UI to render the chart.
+     *
+     * @param output the target to write the JSON to
+     */
+    void writeJSON(JSONStructuredOutput output);
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/dashboards/Chart.java:23]: Would be more fitting to our general naming scheme when writeJSON would be named writeJson, wouldn't it?

File: src/main/java/sirius/biz/tycho/dashboards/EntitiesChartProvider.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+     */
+    protected abstract void collectCharts(Consumer<MetricDescription> descriptionConsumer);
+
+    @Override
+    public Chart resolveChart(String targetName, String chartName) {
+        return resolveChart(chartName);
+    }
+
+    /**
+     * Resolves the given chart.
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/dashboards/EntitiesChartProvider.java:46]: Could be a bit more descriptive :)

File: src/main/java/sirius/biz/tycho/dashboards/EntitiesKeyMetricProvider.java [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+     */
+    protected abstract void collectKeyMetrics(Consumer<MetricDescription> descriptionConsumer);
+
+    @Override
+    public KeyMetric resolveKeyMetric(String targetName, String metricName) {
+        return resolveKeyMetric(metricName);
+    }
+
+    /**
+     * Resolves the given key metric.
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/dashboards/EntitiesKeyMetricProvider.java:48]: Could be a bit more descriptive :)

File: src/main/java/sirius/biz/tycho/dashboards/EntitiesKeyMetricProvider.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+package sirius.biz.tycho.dashboards;
+
+import sirius.db.mixing.Entity;
+import sirius.db.mixing.Mixing;
+
+import java.util.function.Consumer;
+
+/**
+ * Provides a base implementation which provides key metrics for an entity class.
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/dashboards/EntitiesKeyMetricProvider.java:17]: Could maybe describe what a key metric actually is. Or link to a place where it is described.

File: src/main/java/sirius/biz/tycho/dashboards/KeyMetric.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+import sirius.kernel.commons.NumberFormat;
+import sirius.web.services.JSONStructuredOutput;
+
+import javax.annotation.Nullable;
+import java.time.LocalDate;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Represents a key metric to be shown on a metric dashboard.
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/dashboards/KeyMetric.java:22]: Maybe expand this description a bit and link to it from the key metric providers, etc via a `@see`

File: src/main/java/sirius/biz/tycho/dashboards/KeyMetric.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+        this.history = query.valuesUntil(LocalDate.now(), query.determineDefaultLimit());
+        return this;
+    }
+
+    /**
+     * Outputs the key metric as JSON.
+     *
+     * @param output the output to write the JSON to
+     */
+    public void writeJSON(JSONStructuredOutput output) {
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/dashboards/KeyMetric.java:76]: s.a.

File: src/main/java/sirius/biz/tycho/dashboards/KeyMetric.java [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        output.beginObject("metric");
+        output.property("value", value);
+        output.beginArray("history");
+        this.history.forEach(historyValue -> output.property("value", historyValue));
+        output.endArray();
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/dashboards/KeyMetric.java:81]: think there is actually a output.array overload that does this.

File: src/main/java/sirius/biz/tycho/dashboards/MetricDescription.java [hunk 15]
```diff
@@ -0,0 +1,75 @@
...
+ *
+ * Copyright by scireum GmbH
+ * http://www.scireum.de - info@scireum.de
+ */
+
+package sirius.biz.tycho.dashboards;
+
+import sirius.kernel.di.std.Priorized;
+
+public class MetricDescription {
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/dashboards/MetricDescription.java:13]: javadoc?

File: src/main/java/sirius/biz/tycho/dashboards/MetricDescription.java [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+
+    protected String providerName;
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/dashboards/MetricDescription.java:15]: provider name can not be set?

File: src/main/java/sirius/biz/tycho/dashboards/MetricDescription.java [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+    protected String targetName;
```
Comment (sabieber) [inline src/main/java/sirius/biz/tycho/dashboards/MetricDescription.java:16]: target name can not be set other than `-`?

File: src/main/resources/default/assets/scripts/metrics-dashboards.js [hunk 18]
```diff
@@ -0,0 +1,40 @@
...
+        fetch('/tycho/metrics/api', {
+            method: "post",
+            body: JSON.stringify({tasks: tasks})
+        }).then(function (response) {
+            return response.json();
+        }).then(function (response) {
+            for (let i = 0; i < response.tasks.length; i++) {
+                const _element = elements[i];
+                const task = tasks[i];
+                const data = response.tasks[i];
```
Comment (sabieber) [inline src/main/resources/default/assets/scripts/metrics-dashboards.js:25]:
```suggestion
        }).then(function (json) {
            for (let i = 0; i < json.tasks.length; i++) {
                const _element = elements[i];
                const task = tasks[i];
                const data = json.tasks[i];
```

**Review by sabieber — CHANGES_REQUESTED** [review]

File: src/main/java/sirius/biz/tycho/dashboards/BasicChartProvider.java (hunk 6 above)
Comment (andyHa) [reply src/main/java/sirius/biz/tycho/dashboards/BasicChartProvider.java:32]: hasPermission takes a varargs param, we have a set here.... so this seems actually simpler...

File: src/main/java/sirius/biz/tycho/dashboards/BasicKeyMetricProvider.java (hunk 7 above)
Comment (andyHa) [reply src/main/java/sirius/biz/tycho/dashboards/BasicKeyMetricProvider.java:37]: s.a. ^^

File: src/main/java/sirius/biz/tycho/dashboards/BaseTimeseriesChart.java (hunk 4 above)
Comment (sabieber) [reply src/main/java/sirius/biz/tycho/dashboards/BaseTimeseriesChart.java:26]: Yes but the current implementation allows to overwrite the used implementation of the labels List, so someone could pass an immutable list or another one a mutable ArrayList.

Wouldn't it follow best practices when the actual List implementation used was defined here and only the entries of the passed in list were copied in `withLabels` via `addAll`?

**Review by mkeckmkeck — APPROVED** [review]

**Review by sabieber — APPROVED** [review]

**Pull request merged**